<h3>To accurately predict Insurance costs :- </h3>
(for given dataset)

<hr>
<h2><center><b>Importing Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

<hr>
<h2><center><b>Loading Dataset

In [2]:
df=pd.read_csv("Medical_Cost_Prediction.csv")
df

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

<hr>
<h2><center><b>Data Preprocessing

In [3]:
df["sex"]=df["sex"].replace({"female":0,"male":1})
df["smoker"]=df["smoker"].replace({"no":0,"yes":1})

In [4]:
x=df.iloc[:,:-1]
y=df["charges"].values

In [5]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
col_trans=ColumnTransformer(transformers=[('encoder',OneHotEncoder(),[5])],remainder='passthrough')
x=np.array(col_trans.fit_transform(x))

In [6]:
y=y.reshape((-1,1))

In [7]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1)

<hr>
<h2><center><b>Machine Learning Models</h2></center<br>
<h2>Multiple Linear Regression

In [8]:
from sklearn.linear_model import LinearRegression
reg=LinearRegression()
reg.fit(x_train,y_train)
y_pred=reg.predict(x_test)

In [9]:
from sklearn.metrics import r2_score
print("Result of Multiple Linear Regression Model on data is : ",r2_score(y_test,y_pred))

Result of Multiple Linear Regression Model on data is :  0.7623311844057113


# SVR Model

In [16]:
from sklearn.preprocessing import StandardScaler
ss=StandardScaler()
x_std=ss.fit_transform(x_train)
x_test_std=ss.transform(x_test)

from sklearn.svm import SVR
for i in ["poly","sigmoid","rbf","linear"]:
    reg_svr=SVR(kernel=i)
    reg_svr.fit(x_std,y_train.ravel())
    y_pred=reg_svr.predict(x_test_std)
    print("####",i,"\nResult of SVR Model on data is : ",r2_score(y_test,y_pred))

#### poly 
Result of SVR Model on data is :  -0.1112460280096188
#### sigmoid 
Result of SVR Model on data is :  -0.10691599199158563
#### rbf 
Result of SVR Model on data is :  -0.1125411127457776
#### linear 
Result of SVR Model on data is :  -0.028928982185404628


# Decision Tree

In [19]:
from sklearn.tree import DecisionTreeRegressor
reg_tree=DecisionTreeRegressor(random_state=1)
reg_tree.fit(x_std,y_train)
y_pred_d_tree=reg_tree.predict(x_test_std)
print("Result of Decision tree Model on data is : ",r2_score(y_test,y_pred_d_tree))

Result of Decision tree Model on data is :  0.7514843118418003


# Random Forest Reg

In [21]:
from sklearn.ensemble import RandomForestRegressor
reg_ran=RandomForestRegressor(n_estimators=300,random_state=0)
reg_ran.fit(x_std,y_train.ravel())
y_pred_r_forest=reg_ran.predict(x_test_std)
print("Result of Random Forest Model on data is : ",r2_score(y_test,y_pred_r_forest))

Result of Random Forest Model on data is :  0.857011681960076



<h4>Thus it can be seen that "Random Forest Regression Model" has the best fit to out dataset :</h4>
<b>i.e R square value = 0.857</b>

<h2><b>Artificial Neural Network 

In [22]:
import tensorflow as tf

In [23]:
ann = tf.keras.models.Sequential()

In [24]:
ann.add(tf.keras.layers.Dense(units=9,activation="relu"))
ann.add(tf.keras.layers.Dense(units=9,activation="relu"))

In [25]:
ann.add(tf.keras.layers.Dense(units=1))

In [30]:
ann.compile(optimizer="adam",loss="mean_squared_error")
ann.fit(x_std,y_train,epochs=500)

Epoch 1/500
34/34 [==============================] - 1s 3ms/step - loss: 275125472.0000
Epoch 2/500
34/34 [==============================] - 0s 3ms/step - loss: 273352736.0000
Epoch 3/500
34/34 [==============================] - 0s 3ms/step - loss: 271515552.0000
Epoch 4/500
34/34 [==============================] - 0s 3ms/step - loss: 269579488.0000
Epoch 5/500
34/34 [==============================] - 0s 2ms/step - loss: 267573168.0000
Epoch 6/500
34/34 [==============================] - 0s 4ms/step - loss: 265497632.0000
Epoch 7/500
34/34 [==============================] - 0s 3ms/step - loss: 263373376.0000
Epoch 8/500
34/34 [==============================] - 0s 3ms/step - loss: 261153856.0000
Epoch 9/500
34/34 [==============================] - 0s 4ms/step - loss: 258891152.0000
Epoch 10/500
34/34 [==============================] - 0s 2ms/step - loss: 256548768.0000
Epoch 11/500
34/34 [==============================] - 0s 2ms/step - loss: 254155680.0000
Epoch 12/500
34/34 [==========

In [31]:
y_pred=ann.predict(x_test_std)
print(r2_score(y_test,y_pred))

9/9 [==============================] - 0s 1ms/step
0.7941245170401314


<hr>
<h4><center><b><i>Highest R2 Score of 0.857 is achieved by Random Forest Model &#10031;